### PR TITLE
Allow configuration of environment system proxy

### PR DESCRIPTION
**Pull request by GabrielNagy**
The actual application of the environment proxy settings is quite simple, the more interesting part is the abstraction of the proxy settings so we can easily apply them to other backends in the future (like APT and GSettings).

We validate all proxy settings to ensure they're in the right format, we properly escape them while ensuring we don't escape the already escaped ones, then we can represent them in the required format which in this case is as environment variables.

DEENG-553

### CODE REVIEW

File: internal/proxy/setting.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+		proxies = append(proxies, *setting)
+	}
+
+	if noproxy != "" {
+		setting, err := newSetting(noProxy, noproxy)
+		if err != nil {
+			return nil, err
+		}
+		proxies = append(proxies, *setting)
+	}
```
Comment (denisonbarbosa) [inline internal/proxy/setting.go:86]: In order to avoid this, maybe we can change the function args to be a variadic of "pairs" e.g. (protocol, url)?
Then we could also have a "global" map with the protocols as well, e.g.:
```Go
protocols := map[string]protocol {
"http": httpProxy,
...
}
```
It would require quite a few code changes to propagate this behavior up the chain, but maybe this would help avoiding having calls with too many empty strings in the process and ease our life should the number of supported protocols increase (I'm not even sure if this is possible, tbh).
This is more of a rambling and shared thinking rather than a suggestion, so let me know what you think about it.

File: internal/proxy/setting.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+// useGlobalHTTPProxy returns true if only HTTP proxy is set, signaling global
+// use of the HTTP proxy.
+func useGlobalHTTPProxy(http, https, ftp, socks string) bool {
+	if http != "" && https == "" && ftp == "" && socks == "" {
+		return true
+	}
+	return false
+}
+
+func newSetting(proto protocol, uri string) (p *proxySetting, err error) {
```
Comment (denisonbarbosa) [inline internal/proxy/setting.go:100]: Missing doc comment

File: internal/proxy/setting.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+func escapeURLUsername(uri string) string {
+	// Attempt to unescape the string first, discarding any error
+	// At best, this prevents us from escaping the URL multiple times
+	// At worst, the URL is not affected (we will treat % signs as part of the
+	// credentials and escape them later)
+	uri, _ = url.PathUnescape(uri)
+
+	// Regexp to check if the URI contains credentials
+	r := regexp.MustCompile(`^\w+://(?:(?P<credentials>.*:?.*)@)[a-zA-Z0-9.-]+(:[0-9]+)?/?$`)
```
[the remark this answers is not in the document]
Comment (denisonbarbosa) [reply internal/proxy/setting.go:143]: Ack, thanks for the explanation!

File: internal/testutils/golden.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
+package testutils
```
Comment (denisonbarbosa) [inline internal/testutils/golden.go:1]: Yeah, we really should make this a standalone package...

**Review by denisonbarbosa — COMMENTED** [review]
Looks very good to me! I had a few comments and a rambling section, but nothing major. Well done!

File: internal/proxy/proxy_test.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+		})
+	}
+}
+
+func TestApply(t *testing.T) {
+	t.Parallel()
+
+	envConfigPath := filepath.Join("etc", "environment.d", "99ubuntu-proxy.conf")
+
+	initialTime := time.Date(2001, 1, 1, 1, 1, 1, 1, time.UTC)
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:64]: or `Time.Unix(0)` or your birthday date :)

More seriously, we can stat the file after copying it, wdyt?

File: internal/proxy/proxy_test.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		https   string
+		ftp     string
+		socks   string
+		noProxy string
+		mode    string
+
+		existingDirs       []string
+		existingPerms      map[string]os.FileMode
+		prevContents       map[string]string
+		wantUnchangedFiles []string
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:77]: this is not a test behaviour modifier but an expectation, so should be grouped with wantErr, no?

File: internal/proxy/proxy_test.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+		wantErr bool
+	}{
+		"No options set":   {},
+		"HTTP option set":  {http: "http://example.com:8080"},
+		"Some options set": {http: "http://example.com:8080", https: "https://example.com:8080"},
+		"Some options set, environment parent directories are created": {
+			http: "http://example.com:8080", https: "https://example.com:8080", existingDirs: []string{},
+		},
+		"Some options set, environment file is unchanged": {
+			wantUnchangedFiles: []string{envConfigPath},
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:89]: ditto here, I would really put the Want at the end.

File: internal/proxy/proxy_test.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+		wantErr bool
+	}{
+		"No options set":   {},
+		"HTTP option set":  {http: "http://example.com:8080"},
+		"Some options set": {http: "http://example.com:8080", https: "https://example.com:8080"},
+		"Some options set, environment parent directories are created": {
+			http: "http://example.com:8080", https: "https://example.com:8080", existingDirs: []string{},
+		},
+		"Some options set, environment file is unchanged": {
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:88]: `Same options set, environement file should not be changed`?

File: internal/proxy/proxy_test.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+			http:               "http://example.com:8080", https: "https://example.com:8080",
+			prevContents: map[string]string{envConfigPath: fmt.Sprintf(`%s
+HTTP_PROXY=http://example.com:8080
+http_proxy=http://example.com:8080
+HTTPS_PROXY=https://example.com:8080
+https_proxy=https://example.com:8080
+`, proxy.ConfHeader)},
+		},
+		"All options set":                              {http: "http://example.com:8080", https: "https://example.com:8080", ftp: "ftp://example.com:8080", socks: "socks://example.com:8080", noProxy: "localhost,127.0.0.1"},
+		"Password is escaped":                          {http: "http://username:p@$$:w0rd@example.com:8080"},
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:99]: I would just add an empty line here, for the auth use cases.

File: internal/proxy/proxy_test.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+		"Escaped password is not escaped again":        {http: "http://username:p%40$$%3Aw0rd@example.com:8080"},
+		"Domain username is escaped":                   {http: `http://EXAMPLE\bobsmith:p@$$:w0rd@example.com:8080`},
+		"Domain username without password is escaped":  {http: `http://EXAMPLE\bobsmith@example.com:8080`},
+		"Escaped domain username is not escaped again": {http: `http://EXAMPLE%5Cbobsmith@example.com:8080`},
+
+		"Proxy files not created when ran with dry-run": {dryRun: true},
+
+		// Error cases
+		"Error when environment config file is read-only":    {existingPerms: map[string]os.FileMode{envConfigPath: 0444}, prevContents: map[string]string{envConfigPath: "something"}, wantErr: true},
+		"Error when environment directory is a file":         {existingDirs: []string{"etc/"}, prevContents: map[string]string{filepath.Dir(envConfigPath): "something"}, wantErr: true},
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:109]: What you want to trigger here is can’t write to environment directory, so I think the test description should reflect this.

File: internal/proxy/proxy_test.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+		"Error when environment directory is not executable": {existingPerms: map[string]os.FileMode{filepath.Dir(envConfigPath): 0444}, wantErr: true},
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:110]: Similarly, when can’t read directory content?

File: internal/proxy/proxy_test.go [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+				require.NoError(t, err, "Setup: Couldn't change mtime for %q", p)
+			}
+
+			for file, perms := range tc.existingPerms {
+				testutils.Chmod(t, filepath.Join(root, file), perms)
+			}
+
+			ctx := context.Background()
+			if tc.dryRun {
+				ctx = context.WithValue(ctx, proxy.DryRun, true)
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:142]: I would expose DryRun only for tests, wdyt?

File: internal/proxy/proxy_test.go [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+				require.Error(t, err, "Apply should have failed but didn't")
+				return
+			}
+			require.NoError(t, err, "Apply failed but shouldn't have")
+
+			testutils.CompareTreesWithFiltering(t, root, testutils.GoldenPath(t), testutils.Update())
+			for _, file := range tc.wantUnchangedFiles {
+				fi, err := os.Stat(filepath.Join(root, file))
+				require.NoError(t, err, "Setup: Failed to stat proxy config file")
+				require.Equal(t, initialTime, fi.ModTime().UTC(), "Proxy config file was modified")
```
Comment (didrocks) [inline internal/proxy/proxy_test.go:159]: Here, I would more write what we expect from Equal, like "Proxy config file should not be modified"

File: internal/proxy/testdata/TestApply/golden/no_options_set/etc/environment.d/99ubuntu-proxy.conf [hunk 14]
```diff
@@ -0,0 +1 @@
+### This file was generated by ubuntu-proxy-manager - manual changes will be overwritten
```
Comment (didrocks) [inline internal/proxy/testdata/TestApply/golden/no_options_set/etc/environment.d/99ubuntu-proxy.conf:1]: I would change the logic with "no options set" -> no file (and thus, removing file if a previous was around). Then, we have 2 use test cases:
- one with the file around before deleting it
- one with new file and still none after applying "nothing"

**Review by didrocks — CHANGES_REQUESTED** [review]
Some comments in general and suggestions for naming convention and such, but in general, very good work! Let me know what you think about them.

Great to see good progress and high quality here :)

File: internal/proxy/env.go [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+		// #nosec G301 - /etc/environment.d permissions are 0755, so we should keep the same pattern
+		if err := os.MkdirAll(envConfigDir, 0755); err != nil {
+			return fmt.Errorf("failed to create environment config parent directory: %w", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("unexpected error while checking environment directory: %w", err)
+	}
+
+	// #nosec G306 - /etc/environment.d/* permissions are 0644, so we should keep the same pattern
+	if err := os.WriteFile(p.envConfigPath, []byte(content), 0644); err != nil {
```
[the remark this answers is not in the document]
Comment (didrocks) [reply internal/proxy/env.go:55]: I guess if you ask that could be a Friday topic, but yes, we do need it :)
TLDR: file system crash

File: internal/proxy/proxy_test.go (hunk 5 above)
Comment (didrocks) [reply internal/proxy/proxy_test.go:64]: The resolution is nanosecond based, and at least, on other projects, this works :)

Comment (didrocks) [reply internal/proxy/proxy_test.go:64]: why? You can get current time after your crated your own file within the subtest itself?

Comment (didrocks) [reply internal/proxy/proxy_test.go]: ah, you’re right, we are in Apply(), so we will have multiuple files :)

Ok, my only request then is to take a very identifiable date (unix 0) rather than a random one in 2001 (it was a bad year anyway :p)

File: internal/proxy/env.go [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+	return fmt.Sprintf(`%s_PROXY=%s
+%s_proxy=%s
+`,
+		strings.ToUpper(string(p.protocol)), p.escapedURL,
+		p.protocol, p.escapedURL)
+}
+
+// applyEnvironmentProxy applies the proxy configuration in the form of
+// environment variables set in /etc/environment.d.
+func (p Proxy) applyEnvironmentProxy() (err error) {
```
[the remark this answers is not in the document]
Comment (didrocks) [reply internal/proxy/env.go]: well, what you apply are the environment variables themselves, maybe then `applyEnvironmentVariables`?

File: internal/proxy/proxy.go [hunk 17]
```diff
@@ -3,20 +3,54 @@ package proxy
...
 // New returns a new instance of a proxy manager.
-func New(ctx context.Context, http, https, ftp, socks, no, mode string) (*Proxy, error) {
-	return &Proxy{}, nil
+func New(ctx context.Context, http, https, ftp, socks, no, mode string, args ...option) (p *Proxy, err error) {
+	defer decorate.OnError(&err, "couldn't create proxy manager")
+
+	// Set default options
+	opts := options{
+		root:          "/",
+		envConfigPath: "etc/environment.d/99ubuntu-proxy.conf",
```
[the remark this answers is not in the document]
Comment (didrocks) [reply internal/proxy/proxy.go]: yes, it’s just that ubuntu-proxy seems to be very specific to ubuntu, while we want to should in the filename the source of it. Maybe it could be directly ubuntu-proxy-manager.conf?

File: internal/proxy/setting.go [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/ubuntu/decorate"
+)
+
+type protocol string
+
+const (
```
[the remark this answers is not in the document]
Comment (didrocks) [reply internal/proxy/setting.go:16]: That was ironic btw, the fact it would be great to have that :) The iota is good if you never print it to the end user. :)

File: internal/proxy/proxy_test.go (hunk 12 above)
Comment (didrocks) [reply internal/proxy/proxy_test.go:145]: I have no strong opinion about it. I would prefer to keep dry-run not exposed, but just take whatever path you prefer between the 2 options.

**Review by didrocks — COMMENTED** [review]
And here we go :)

**Review by didrocks — APPROVED** [review]
